### DISCUSSION TITLE
Cannot construct instance of `com.webauthn4j.data.attestation.authenticator.AuthenticatorData`

### DISCUSSION THREAD

**Discussion by srikanthramu** [webauthn4j/webauthn4j#695] [Q&A] [answered]
Hello,

I'll start by acknowledging the work done by the webauthn4j maintainers and contributors. Thank you. 

I am building a Burp extension to support decoding/encoding of cbor content in WebAuthn flow (https://github.com/srikanthramu/webauthn-cbor-burp). For this I am working on deserializing and serializing cbor format in AttestationObject using webauthn4j library. Here is a snippet that results in error:

-----

        String cborStr = // https://github.com/webauthn4j/webauthn4j/blob/master/webauthn4j-test/src/main/java/com/webauthn4j/test/TestDataUtil.java#L115
        // construct attestationobjectcovertor
        ObjectConverter objectConverter = new ObjectConverter();
        AttestationObjectConverter attestationObjectConverter = new AttestationObjectConverter(objectConverter);
        
        // deserialize the attestationobject to json string
        AttestationObject attestObj = attestationObjectConverter.convert(cborStr);
        JsonConverter jsonConverter = objectConverter.getJsonConverter();
        String decodedAttestObjString = jsonConverter.writeValueAsString(attestObj);
        System.out.println( "Decoded CBOR String = " + decodedAttestObjString );
        
        // reserialize json string to base64 cbor format
        attestObj = jsonConverter.readValue(decodedAttestObjString, AttestationObject.class);
        String encodedAttestObjString = attestationObjectConverter.convertToBase64urlString(attestObj);
        System.out.println( "new CBOR String = " + encodedAttestObjString);
--------
Error: "Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `com.webauthn4j.data.attestation.authenticator.AuthenticatorData` (no Creators, like default constructor, exist): cannot deserialize from Object value (no delegate- or property-based Creator)". 

Any pointers is appreciated?

**Comment by ynojima** [comment, marked as answer]
`JsonConverter` is not designed to serialize/deserialize `AttestationObject`.
Please use `AttestationObjectConverter` instead for serializatrion.

**Reply by srikanthramu** [reply]
Thanks for the quick response, yes I do plan to use AttestationObjectConverter (please refer to my snippet where I've used AttestationObjectConverter to first convert the base64 url encoded cbor format to attestation object) but I would need the attestationobject in JSON string format hence used JsonConverter.  AttestationObject toString() doesn't return in json format either so I was planning to JsonConverter.

**Reply by ynojima** [reply]
If you just want to serialize `AttestationObject` into string, use `AttestationObjectConverter`. It will convert it to WebAuthn standard byte array representation.
Do you indeed want to serialize it to JSON string? Could you tell me the reason? To get a human-readable representation used in debugging purpose?

**Reply by srikanthramu** [reply]
Yes, to get a human readable format. For example: https://github.com/srikanthramu/webauthn-cbor-burp/blob/main/images/webauthn-cbor-decode.png

**Reply by ynojima** [reply]
I see, but why do you want to deserialize from JSON?

**Reply by srikanthramu** [reply]
I am not sure if I am following your question, so the goal for the Burp extension is to give users the ability to intercept the request containing the attestionobject (in CBOR format) and to modify the request and send the modified attestationobject to the server. My plan is to first decode the attestation object to json string so that the users can modify the request for example, change the "fmt" attribute to "none" and then convert the json string to the original cbor format and send the request. Hence, I think I need to first decode the base64 url encoded cbor data to the attestationObject using the AttestationObjectConvertor and then convert it to json string. If the user changes data, reserialize the modified json string to attestationObject in the form of base64 url encoded cbor data.

**Reply by ynojima** [reply]
I see. So, what you need is a serializer/deserializer that converts WebAuthn object model into a pure JSON string from root to the leaf, but but what WebAuthn4J provides is a converter that serialize in a format defined in WebAuthn spec, which contains CBOR or non self descriptive byte array.
Providing pure JSON string converter is not WebAuthn4J goal, WebAuthn4J library cannot help you directly, but some custom JSON serializer/deserializer can be reused when you implement your own converter with Jackson ObjectMapper.

If I were you,
1. Create a custom Jackson Module based on [SimpleModule](https://fasterxml.github.io/jackson-databind/javadoc/2.4/com/fasterxml/jackson/databind/module/SimpleModule.html)
2. Reuse some JSON [serializers](https://github.com/webauthn4j/webauthn4j/tree/master/webauthn4j-core/src/main/java/com/webauthn4j/converter/jackson/serializer/json)/[deserializers](https://github.com/webauthn4j/webauthn4j/tree/master/webauthn4j-core/src/main/java/com/webauthn4j/converter/jackson/deserializer/json) from WebAuthn4J
3. Implement rest WebAuthn4J data model serializers/deserializers that needs custom converter (ex. AuthenticatorData). When possible, use webauthn4j CBOR serializers/deserializers implementations as reference.
4. Create Jackson ObjectMapper instance and register your custom Jackson Module
5. Use the Jackson ObjectMapper to convert from/to pure JSON string.

**Reply by srikanthramu** [reply]
Thank you for your detailed response
